src/host-services.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  getServicePromiseForDoc,
+  registerServiceBuilderForDoc,
+} from './service';
+
+const ServiceNames = {
+  VISIBILITY: 'host-visibility',
+  FULLSCREEN: 'host-fullscreen',
+  EXIT: 'host-exit',
+};
+
+export class HostServices {
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static visibilityForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Visibility>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.VISIBILITY));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Visibility} impl
+   */
+  static installVisibilityServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.VISIBILITY, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Visibility>}
+   */
+  static fullscreenForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.FULLSCREEN));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Fullscreen} impl
+   */
+  static installFullscreenServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.FULLSCREEN, impl, /* opt_instantiate */ true);
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @return {!Promise<!Exit>}
+   */
+  static exitForDoc(elementOrAmpDoc) {
+    return /** @type {!Promise<!Fullscreen>} */ (
+      getServicePromiseForDoc(elementOrAmpDoc, ServiceNames.EXIT));
+  }
+
+  /**
+   * @param {!Element|!./service/ampdoc-impl.AmpDoc} elementOrAmpDoc
+   * @param {!Exit} impl
+   */
+  static installExitServiceForDoc(elementOrAmpDoc, impl) {
+    registerServiceBuilderForDoc(elementOrAmpDoc,
+        ServiceNames.EXIT, impl, /* opt_instantiate */ true);
+  }
+}
+
+/**
+ * Visibility defines interface provided by host for visibility detection.
+ *
+ * @interface
+ */
+export class Visibility {
+
+  /**
+   * Register a callback for scroll events.
+   * @param {function(!VisibilityDataDef)} unusedCallback
+   */
+  onVisibilityChange(unusedCallback) {}
+}
+
+/**
```

Comment:
In some cases we'll want analytics in AMPHTML to do tricky things based on the presence of mraid.js  Will this architecture support that?  I think what we'll want is something like:

case 1, server side we think we're serving to mApp: use mraid, never use intersection observer
case 2, server side we think we're serving to mWeb: don't even try to use mraid, use intersection observer
case 3, server side we can't tell: if mraid is present report failure, otherwise use intersection observer

What's going on in case 3 is that we can only use the visibility information from mraid if we've tested and verified the client, and we're always able to tell if that's the situation server-side.  Which means if we find ourselves in a mobile app, which we can tell because mraid.js will be present, and we're not certified, then we aren't allowed to use visibility information from MRAID, but we also can't use it from intersection observer because that's not reporting the right thing in a mApp context.

There's a case 4 as well: use mraid if present, otherwise fall back to intersection observer.  This isn't something we want, because of restrictions around when we're allowed to use visibility data, but since it does represent the best synthesis of information we have about visibility it's something I think other people are likely to want.

Reply:
Thinking for a while, the easiest way might be to let each extension handle the fallback themselves.

i.e. if amp-mraid.js see a failure, it register the the InOb implementation instead.